server/src/main/java/org/opensearch/common/xcontent/JsonToStringXContentParser.java

```diff
@@ -72,21 +73,31 @@ public XContentParser parseObject() throws IOException {
         builder.startObject();
         LinkedList<String> path = new LinkedList<>(Collections.singleton(fieldTypeName));
         while (currentToken() != Token.END_OBJECT) {
-            parseToken(path);
+            parseToken(path, null);
         }
-        builder.field(this.fieldTypeName, keyList);
-        builder.field(this.fieldTypeName + VALUE_SUFFIX, valueList);
-        builder.field(this.fieldTypeName + VALUE_AND_PATH_SUFFIX, valueAndPathList);
+        // deduplication the fieldName,valueList,valueAndPathList
+        builder.field(this.fieldTypeName, new HashSet<>(keyList));
+        builder.field(this.fieldTypeName + VALUE_SUFFIX, new HashSet<>(valueList));
+        builder.field(this.fieldTypeName + VALUE_AND_PATH_SUFFIX, new HashSet<>(valueAndPathList));
```

Comment:
Could you do a `LinkedHashSet<>` instead? It would preserve the order, so the unit tests would remain consistent. (Right now, the unit tests are relying on the order of the `HashSet`, which is hard to reason about.)

Reply:
I tried that out and was able to remove most of your unit test diff, just keeping the deduplication of field values: https://github.com/opensearch-project/OpenSearch/compare/main...msfroh:OpenSearch:fix_null_flat_object